sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -496,19 +509,38 @@ case class Overlay(input: Expression, replace: Expression, pos: Expression, len:
     this(str, replace, pos, Literal.create(-1, IntegerType))
   }
 
-  override def dataType: DataType = StringType
+  override def dataType: DataType = input.dataType
 
-  override def inputTypes: Seq[AbstractDataType] =
-    Seq(StringType, StringType, IntegerType, IntegerType)
+  override def inputTypes: Seq[AbstractDataType] = Seq(TypeCollection(StringType, BinaryType),
+    TypeCollection(StringType, BinaryType), IntegerType, IntegerType)
 
   override def children: Seq[Expression] = input :: replace :: pos :: len :: Nil
 
+  override def checkInputDataTypes(): TypeCheckResult = {
```

Comment:
what if the case (string, binary) or (binary, string)?

Reply:
This case not be allowed.

Reply:
It seems its accepted in pg;
```
postgres=# select overlay('aaaaaa' placing '000'::bytea from 2);
    overlay     
----------------
 \x613030306161
(1 row)
```
Not allowed in the other systems and the standard?

Reply:
According ANSI SQL,  only (binary, binary) or (string, string)

Reply:
```
<character overlay function> ::=
OVERLAY <left paren> <character value expression> PLACING <character value expression>
FROM <start position> [ FOR <string length> ]
[ USING <char length units> ] <right paren>
```
and
```
<binary overlay function> ::=
OVERLAY <left paren> <binary value expression> PLACING <binary value expression>
FROM <start position> [ FOR <string length> ] <right paren>
```
I can't find any other case.

Reply:
I think this is an issue about implicit casts, not function arguments.
In the example above, the left argument (text) is casted as binary then `overlay(binary, binary)` is called actually. So, my question is that we need to extend `ImplicitCastInputTypes` for overlay?

Reply:
This is a good idea. I think the behavior of Postgresql's overlay seems a expand of ANSI standard.

Reply:
Just in case, I just want to know the other DBMS behaviours, could you check?

Reply:
As I know,
Vertica and DB2 only support `overlay(string, string)`.
Oracle support `overlay(string, string)`, `overlay(binary, binary)`, `overlay(string, binary)`, `overlay(binary, string)`.

Reply:
I just find fink only support `overlay(string, string)` too.